src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/persistence/FileGC.java

```diff
@@ -0,0 +1,147 @@
+
+  protected Deque<File> eligibleForGc;
+
+  FileGC(
+      Path db_dir,
+      String base_db_filename,
+      TimeUnit time_unit,
+      long periodicity1,
+      int files_count) {
+    this(
+        db_dir, base_db_filename, time_unit, periodicity1, files_count, System.currentTimeMillis());
+  }
+
+  FileGC(
+      Path db_dir,
+      String base_db_filename,
+      TimeUnit time_unit,
+      long periodicity1,
+      int files_count,
+      long currentMillis) {
+    DB_DIR = db_dir;
+    BASE_DB_FILENAME = base_db_filename;
+
+    TIME_UNIT = time_unit;
+    PERIODICITY = periodicity1;
+    FILES_COUNT = files_count;
+
+    List<File> remainingFiles = cleanupAndGetRemaining(currentMillis);
+
+    eligibleForGc = new LinkedList<>(remainingFiles);
+  }
+
+  /**
+   * When we add a newly rotated file, this method checks if the current count of the garbage files
+   * is above the threshold, then it removes the oldest file to keep the count under FILES_COUNT.
+   *
+   * @param filename The name of the new file to be added.
+   */
+  void eligibleForGc(String filename) {
+    File file = Paths.get(DB_DIR.toString(), filename).toFile();
+    if (file.exists()) {
+      eligibleForGc.addLast(file);
+
+      if (eligibleForGc.size() > FILES_COUNT) {
+        File oldestFile = eligibleForGc.removeFirst();
+        delete(oldestFile);
+      }
+    }
+  }
+
+  protected List<File> cleanupAndGetRemaining(long currentMillis) {
+    String[] files = getDbFiles();
+    List<File> afterCleaningOldFiles = timeBasedCleanup(files, currentMillis);
+    return countBasedCleanup(afterCleaningOldFiles);
+  }
+
+  protected String[] getDbFiles() {
+    return DB_DIR
+        .toFile()
+        .list(new WildcardFileFilter(BASE_DB_FILENAME + "." + WILDCARD_CHARACTER));
+  }
+
+  protected List<File> timeBasedCleanup(String[] files, final long currentMillis) {
+    long timeDelta = TimeUnit.MILLISECONDS.convert(PERIODICITY, TIME_UNIT) * (FILES_COUNT + 1);
+    long timeLimit = currentMillis - timeDelta;
+
+    List<File> remainingFiles = new ArrayList<>();
+
+    for (String file : files) {
+      Path filePath = Paths.get(DB_DIR.toString(), file);
+      long lastModified = filePath.toFile().lastModified();
+
+      if (lastModified < timeLimit) {
+        delete(filePath.toFile());
+      } else {
+        remainingFiles.add(filePath.toFile());
+      }
+    }
+    return remainingFiles;
+  }
+
+  protected List<File> countBasedCleanup(List<File> files) {
+    int numToDelete = files.size() - FILES_COUNT;
+    Collections.sort(files, LastModifiedFileComparator.LASTMODIFIED_COMPARATOR);
+
+    int deletedSoFar = 0;
+    List<File> remainingFiles = new ArrayList<>();
+
+    for (File file : files) {
+      if (deletedSoFar < numToDelete) {
+        delete(file);
+        deletedSoFar += 1;
+      } else {
+        remainingFiles.add(file);
+      }
+    }
+    return remainingFiles;
+  }
+
+  private void delete(File file) {
+    if (!file.delete()) {
```

Comment:
Do we need to catch exception here ?

Reply:
Good catch! we should. Added a revision to throw exception if we fail to delete not just silently logging it